flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/OperatorCoordinator.java

```diff
@@ -202,8 +202,7 @@ default void notifyCheckpointAborted(long checkpointId) {}
          * target TaskManager. The future is completed exceptionally if the event cannot be sent.
          * That includes situations where the target task is not running.
          */
-        CompletableFuture<Acknowledge> sendEvent(OperatorEvent evt, int targetSubtask)
-                throws TaskNotRunningException;
+        CompletableFuture<Acknowledge> sendEvent(OperatorEvent evt, int targetSubtask);
```

Comment:
I was wondering here as well a bit.

For the next PR (introducing `SubtaskGateway`), the contract I want to have is the following:
  - The `SubtaskGateway` only throws an Exception before it is ready. That should not really be possible to happen for callers, because they should not have access to the gateway before.
  - Once it was ready, it never throws exceptions, but only completes the future exceptionally. Callers can mostly decide to ignore the result of the future, if they handle the `notifyTaskFaile()` call. If futures fail due to RPC loss, the task will also be failed, so no need to handle the future extra.
  - In the next version of the fix branch, there is no checked exception any more on the subtaskGateway send method.

Reply:
> Callers can mostly decide to ignore the result of the future, if they handle the `notifyTaskFaile()` call.

> a dated SubtaskGateway could throw TaskNotRunningException after subtask failed/reset for earlier failing possible active activities.

With `subtaskReady` and `SubtaskGateway`, even silence should not hurt correctness. I think all should work. But given that future is most likely ignored, I am a bit prefer to call stack exception. It could be more helpful in diagnostics of buggy code.

Reply:
Hmmm, I see value of "received" future. To solely solve FLINK-21996(eg. source split assignment exactly-once), `SourceCoordinator` could take the received future into account for specific events during checkpointing. I am not sure how next pr will look like. But I think this(my assumed approach) actually means that `OperatorCoordinator` authors should also take exactly-once into account if they want this.

Still not sure how this could be helpful for custom source events and no source task.

Reply:
I will leave the "received" future for now, because this is on the more general `OperatorCoordinator` interface. The `SourceEnumerator` interface doesn't expose this, which is the most important part (public API).

Reply:
I was thinking whether the signature is mixed with two different requirements:
* `void transferState(OperatorStateEvent evt) throws TaskNotRunningException`(throwing is optional but might be useful) Sending failure will failover subtask. This actually means that coordinator want to transfer part of its state to subtask.
* ` CompletableFuture<Xyz> sendEvent(OperatorEvent evt, Duration timeout)` Send an event with not exactly-once guarantee. It is author's responsibility to dealing with failure hence the timeout parameter. The event does not contribute or belong to state in either side.